### DISCUSSION TITLE
MemorySegment-based Buffer implementation

### DISCUSSION THREAD

**Discussion by BomBardyGamer** [netty/netty#12905] [Ideas]
I read the original PR that changed the Netty 5 buffer API the other day, and noticed that the original implementation was based on the Foreign Function & Memory API, specifically `MemorySegment`, however this implementation was dropped due to the indeterminate state of the API at the time. As it is now in preview in JDK 19, and likely to become release within the next few releases, are there any plans to provide that implementation? Whether in a separate module, an incubator module, or something else (I don't think it's ready enough to be mainstream yet, especially as it's very new and Netty depends on a much older version of Java), as I believe the implementation could be beneficial, seeing the current benefits of the current Foreign Function & Memory API.

Just an idea. I don't know too much about the Foreign Function & Memory API, but from what I've seen, it seems like a good idea.

**Comment by BomBardyGamer** [comment]
After further investigation, it seems like the old implementation was left in the [original incubator API project](https://github.com/netty/netty-incubator-buffer-api), however it is unclear what the state of that project is, or the state of the MemorySegment-based implementation in that project. If this could be clarified, it would be much appreciated.

**Comment by chrisvest** [comment]
Plan is to move it into the main Netty repository, but we need to think about what to do wrt. releasing, since we're not keen on depending on preview features, as they change from release to release. Our MemorySegment integration is currently tracking JDK 20 snapshot.
